crates/astria-sequencer/src/service/mempool/mod.rs

```diff
@@ -120,50 +127,181 @@ async fn handle_check_tx<S: accounts::StateReadExt + address::StateReadExt + 'st
+    // check if the transaction has been removed from the appside mempool and handle
+    // the removal reason
+    if let Some(removal_reason) = mempool.check_removed_comet_bft(tx_hash).await {
+        match removal_reason {
+            RemovalReason::Expired => {
+                metrics.increment_check_tx_removed_expired();
+                return Some(response::CheckTx {
+                    code: Code::Err(AbciErrorCode::TRANSACTION_EXPIRED.value()),
+                    info: "transaction expired in app's mempool".into(),
+                    log: "Transaction expired in the app's mempool".into(),
+                    ..response::CheckTx::default()
+                });
+            }
+            RemovalReason::FailedPrepareProposal(err) => {
+                metrics.increment_check_tx_removed_failed_execution();
+                return Some(response::CheckTx {
+                    code: Code::Err(AbciErrorCode::TRANSACTION_FAILED.value()),
+                    info: "transaction failed execution in prepare_proposal()".into(),
+                    log: format!("transaction failed execution because: {err}"),
+                    ..response::CheckTx::default()
+                });
+            }
+            RemovalReason::NonceStale => {
+                return Some(response::CheckTx {
+                    code: Code::Err(AbciErrorCode::INVALID_NONCE.value()),
+                    info: "transaction removed from app mempool due to stale nonce".into(),
+                    log: "Transaction from app mempool due to stale nonce".into(),
+                    ..response::CheckTx::default()
+                });
+            }
+            RemovalReason::LowerNonceInvalidated => {
+                return Some(response::CheckTx {
+                    code: Code::Err(AbciErrorCode::LOWER_NONCE_INVALIDATED.value()),
+                    info: "transaction removed from app mempool due to lower nonce being \
+                           invalidated"
+                        .into(),
+                    log: "Transaction removed from app mempool due to lower nonce being \
+                          invalidated"
+                        .into(),
+                    ..response::CheckTx::default()
+                });
+            }
+        }
+    };
+
+    let finished_removal_check = Instant::now();
+    metrics.record_check_tx_duration_seconds_check_removed(
+        finished_removal_check.saturating_duration_since(start_removal_check),
+    );
+
+    None
+}
+
+/// Stateless checks return a [`response::CheckTx`] if the transaction fails any of the checks.
+/// Otherwise, it returns the [`SignedTransaction`] to be inserted into the mempool.
+enum StatelessReturn {
+    Response(response::CheckTx),
+    Tx(SignedTransaction),
+}
+
+/// Performs stateless checks on the transaction.
+///
+/// Returns a [`response::CheckTx`] if the transaction fails any of the checks.
+/// Otherwise, it returns the [`SignedTransaction`] to be inserted into the mempool.
```

Comment:
you can use a `Result<SignedTransaction, response::CheckTx>` as this function's return